sql/src/main/java/io/crate/analyze/where/DocKeys.java

```diff
@@ -0,0 +1,174 @@
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+import io.crate.planner.symbol.Literal;
+import io.crate.planner.symbol.Symbol;
+import io.crate.planner.symbol.ValueSymbolVisitor;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.Base64;
+import org.elasticsearch.common.Preconditions;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+
+public class DocKeys implements Iterable<DocKeys.DocKey> {
+
+    private final int width;
+    private int clusteredByIdx;
+    private final boolean withVersions;
+    private final List<List<Symbol>> docKeys;
+    private final List<Integer> partitionIdx;
+
+    public class DocKey {
+
+        private final List<Symbol> key;
+        private String id;
+
+        public DocKey(int pos) {
+            key = docKeys.get(pos);
+        }
+
+        public Optional<Long> version() {
+            if (withVersions && key.get(width) != null) {
+                return Optional.of((Long) ((Literal) key.get(width)).value());
+            }
+            return Optional.absent();
+        }
+
+        public String routing() {
+            if (clusteredByIdx >= 0) {
+                return ValueSymbolVisitor.STRING.process(key.get(clusteredByIdx));
+            } else {
+                return id();
+            }
+
+        }
+
+        public Optional<List<BytesRef>> partitionValues() {
+            if (partitionIdx == null || partitionIdx.isEmpty()) {
+                return Optional.absent();
+            }
+            List<BytesRef> values = Lists.transform(partitionIdx, new Function<Integer, BytesRef>() {
+                @Nullable
+                @Override
+                public BytesRef apply(Integer input) {
+                    return ValueSymbolVisitor.BYTES_REF.process(key.get(input));
+                }
+            });
+            return Optional.of(values);
+        }
+
+        public String id() {
+            if (id == null) {
+                if (width == 1) {
+                    id = ValueSymbolVisitor.STRING.process(key.get(0));
+                } else {
+                    BytesStreamOutput out = new BytesStreamOutput(width * 16);
+                    try {
+                        out.writeVInt(width);
+                        if (clusteredByIdx >= 0) {
+                            out.writeBytesRef(ValueSymbolVisitor.BYTES_REF.process(key.get(clusteredByIdx)));
+                        }
+                        for (int i = 0; i < width; i++) {
+                            if (i != clusteredByIdx) {
+                                out.writeBytesRef(ValueSymbolVisitor.BYTES_REF.process(key.get(i)));
+                            }
+                        }
+                        out.close();
+                    } catch (IOException e) {
+                        //
```

Comment:
maybe bubble the Exception as RuntimeException using ``Throwables.propagate()`` ?